### PR TITLE
bpo-37957: Allow regrtest to receive a file with test (and subtests) to ignore

### PR DESCRIPTION
**Pull request by pablogsal**
https://bugs.python.org/issue37957

### CODE REVIEW

File: Lib/test/support/__init__.py [hunk 1]
```diff
@@ -2049,7 +2049,9 @@ def _run_suite(suite):
 
 # By default, don't filter tests
 _match_test_func = None
-_match_test_patterns = None
+
+_accept_test_patterns = None
+_ignore_test_patterns = None
```
Comment (vstinner) [inline Lib/test/support/__init__.py:2054]: nitpick: i would prefer to use "ignore" everywhere, or "reject" everywhere, but not both.

File: Lib/test/support/__init__.py [hunk 2]
```diff
@@ -2070,21 +2072,48 @@ def _is_full_match_test(pattern):
     return ('.' in pattern) and (not re.search(r'[?*\[\]]', pattern))
 
 
-def set_match_tests(patterns):
-    global _match_test_func, _match_test_patterns
+def set_match_tests(accept_patterns, reject_patterns):
```
Comment (vstinner) [inline Lib/test/support/__init__.py]: Maybe use None as default value to "document" that None is accepted:

```suggestion
 def set_match_tests(accept_patterns=None, reject_patterns=None):
```

File: Lib/test/support/__init__.py [hunk 3]
```diff
@@ -2070,21 +2072,48 @@ def _is_full_match_test(pattern):
...
+            if accept_func:
+                accept = accept_func(test_id)
+            if reject_func:
+                reject = reject_func(test_id)
+            return accept and not reject
+
+        _match_test_func = match_function
+
+
+def _update_match_function(patterns):
```
Comment (vstinner) [inline Lib/test/support/__init__.py]: I'm fine with "_update_match_function" name. Another propositions: "_compile_match_function" or "_create_match_function".

File: Lib/test/support/__init__.py [hunk 4]
```diff
@@ -2070,21 +2072,48 @@ def _is_full_match_test(pattern):
...
+    accept_func = reject_func = None
+
+    if accept_patterns != _accept_test_patterns:
+        accept_patterns, accept_func = _update_match_function(accept_patterns)
+    if reject_patterns != _ignore_test_patterns:
+        reject_patterns, reject_func = _update_match_function(reject_patterns)
+
+    # Create a copy since patterns can be mutable and so modified later
+    _accept_test_patterns = tuple(accept_patterns)
+    _ignore_test_patterns = tuple(reject_patterns)
```
Comment (vstinner) [inline Lib/test/support/__init__.py]: Maybe do that in _update_match_function()?

File: Lib/test/support/__init__.py [hunk 5]
```diff
@@ -2070,21 +2072,48 @@ def _is_full_match_test(pattern):
...
+
+    if accept_func or reject_func:
```
Comment (vstinner) [inline Lib/test/support/__init__.py]: I prefer to explicitly check for None:

```suggestion
    if accept_func is not None or reject_func is not None:
```

File: Lib/test/test_regrtest.py [hunk 6]
```diff
@@ -961,6 +979,47 @@ def parse_methods(self, output):
...
+                    pass
+        """)
+        all_methods = ['test_method1', 'test_method2',
+                       'test_method3', 'test_method4']
+        testname = self.create_test(code=code)
+
+        # by default, all methods should be run
+        output = self.run_tests("-v", testname)
+        methods = self.parse_methods(output)
+        self.assertEqual(methods, all_methods)
```
Comment (vstinner) [inline Lib/test/test_regrtest.py]: I don't think that this test is useful, but you can keep it if you prefer :-)

**Review by vstinner — COMMENTED** [review]
I love this feature :-) Here is my review.

File: Lib/test/test_regrtest.py (hunk 6 above)
Comment (pablogsal) [reply Lib/test/test_regrtest.py]: I can remove it :)

Comment (pablogsal) [reply Lib/test/test_regrtest.py]: Do you mean the whole `test_ignorefile` test or just that assert?

File: Lib/test/support/__init__.py (hunk 4 above)
Comment (pablogsal) [reply Lib/test/support/__init__.py]: Hummmm, I would suggest to have `_update_match_function` still be agnostic of the pattern is generating and be free of side effects

File: Lib/test/test_regrtest.py (hunk 6 above)
Comment (vstinner) [reply Lib/test/test_regrtest.py]: My comment is on these 4 lines:

```
        # by default, all methods should be run
        output = self.run_tests("-v", testname)
        methods = self.parse_methods(output)
        self.assertEqual(methods, all_methods)
```

File: Lib/test/support/__init__.py [hunk 7]
```diff
@@ -2065,26 +2067,53 @@ def _is_full_match_test(pattern):
...
+def _compile_match_function(patterns):
     if not patterns:
         func = None
         # set_match_tests(None) behaves as set_match_tests(())
         patterns = ()
     elif all(map(_is_full_match_test, patterns)):
         # Simple case: all patterns are full test identifier.
         # The test.bisect_cmd utility only uses such full test identifiers.
-        func = set(patterns).__contains__
+        func = lambda elem: elem in set(patterns)
```
Comment (vstinner) [inline Lib/test/support/__init__.py]: Hum, you create a new temporary set object at each call. `set(patterns).__contains__` creates the set exactly once. No?

File: Lib/test/test_support.py [hunk 8]
```diff
@@ -527,52 +527,100 @@ def id(self):
         test_access = Test('test.test_os.FileTests.test_access')
         test_chdir = Test('test.test_os.Win32ErrorTests.test_chdir')
 
+        # Test acceptance
         with support.swap_attr(support, '_match_test_func', None):
             # match all
-            support.set_match_tests([])
+            support.set_match_tests([], None)
```
Comment (vstinner) [inline Lib/test/test_support.py]: You might omit None here now. It's up to you.

File: Lib/test/test_support.py [hunk 9]
```diff
@@ -527,52 +527,100 @@ def id(self):
...
 
-            support.set_match_tests(['test_access', 'DONTMATCH'])
+            support.set_match_tests(['test_access', 'DONTMATCH'], None)
             self.assertTrue(support.match_test(test_access))
             self.assertFalse(support.match_test(test_chdir))
 
+        # Test rejection
+        with support.swap_attr(support, '_match_test_func', None):
+            # match all
+            support.set_match_tests(None, [])
```
Comment (vstinner) [inline Lib/test/test_support.py]: You might write support.set_match_tests(ignore_patterns=[]) instead. It's up to you.

File: Lib/test/support/__init__.py (hunk 7 above)
Comment (pablogsal) [reply Lib/test/support/__init__.py]: Indeed, good catch!

File: Lib/test/test_regrtest.py (hunk 6 above)
Comment (pablogsal) [reply Lib/test/test_regrtest.py]: I have removed them :)

**Review by vstinner — APPROVED** [review]
LGTM, thanks for the update.

**Comment by pablogsal** [comment]
Thanks for the review, @vstinner  :tada:

**Pull request merged**